server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -589,6 +612,158 @@ void recoverFilesFromSourceAndSnapshot(ShardRecoveryPlan shardRecoveryPlan,
+                    }
+                };
+                requestFuture.addListener(sendRequestListener);
+
+                trackOutstandingRequest(requestFuture);
+                recoveryTarget.restoreFileFromSnapshot(
+                    snapshotFilesToRecover.getRepository(),
+                    snapshotFilesToRecover.getIndexId(),
+                    snapshotFileToRecover,
+                    ActionListener.runBefore(requestFuture, () -> unTrackOutstandingRequest(requestFuture))
+                );
+            } catch (CancellableThreads.ExecutionCancelledException e) {
+                onCancellation(e);
+            } catch (Exception e) {
+                unTrackOutstandingRequest(requestFuture);
+                onRequestCompletion(snapshotFileToRecover.metadata(), e);
+            }
+        }
+
+        void onCancellation(Exception e) {
+            if (cancelled.compareAndSet(false, true)) {
+                pendingSnapshotFilesToRecover.clear();
+                notifyFailureOnceAllOutstandingRequestAreDone(e);
+            }
+        }
+
+        void onRequestCompletion(StoreFileMetadata storeFileMetadata, @Nullable Exception exception) {
+            if (cancelled.get()) {
+                return;
+            }
+
+            if (exception != null) {
+                addFileFailedToRecoverFromSnapshot(storeFileMetadata);
+            }
+
+            if (countDown.countDown()) {
+                final List<StoreFileMetadata> failedToRecoverFromSnapshotFiles = getFilesFailedToRecoverFromSnapshot();
+                listener.onResponse(failedToRecoverFromSnapshotFiles);
+            } else {
+                sendRequest();
+            }
+        }
+
+        synchronized void addFileFailedToRecoverFromSnapshot(StoreFileMetadata storeFileMetadata) {
+            if (filesFailedToDownloadFromSnapshot == null) {
+                filesFailedToDownloadFromSnapshot = new ArrayList<>();
+            }
+            filesFailedToDownloadFromSnapshot.add(storeFileMetadata);
+        }
+
+        synchronized List<StoreFileMetadata> getFilesFailedToRecoverFromSnapshot() {
+            return Objects.requireNonNullElse(filesFailedToDownloadFromSnapshot, Collections.emptyList());
+        }
+
+        private void trackOutstandingRequest(ListenableFuture<Void> future) {
+            synchronized (outstandingRequests) {
+                outstandingRequests.add(future);
+            }
+        }
+
+        private void unTrackOutstandingRequest(ListenableFuture<Void> future) {
+            synchronized (outstandingRequests) {
+                outstandingRequests.remove(future);
+            }
+        }
+
+        private void notifyFailureOnceAllOutstandingRequestAreDone(Exception e) {
+            assert cancelled.get();
+
+            final Set<ListenableFuture<Void>> pendingRequests;
+            synchronized (outstandingRequests) {
+                pendingRequests = outstandingRequests;
```

Comment:
This looks unsafe to me. If we have two requests coming through at the same time:

1. R1 checks for cancelled, which is false. Let us assume it pauses for a while.
2. R2 checks for cancelled, is true. It marks cancelled and goes here to copy the reference to the hash set.
3. R1 resumes and adds its future to the `outstandingRequests`.
4. R2 continues but since there is no synchronization with R1 it could be looking at a garbled hash set.

But even with more synchronization here, I think we could risk missing waiting for R1.

I think checking for `cancellableThreads.isCancelled` inside the mutex in `trackOutstadingRequests` will solve it. Then if true, simply do `checkCancelled` outside the mutex to have it throw.

Additionally, `untrackOutstandingRequests` could also do an unsafe change of the hash-set. We should therefore preventively copy the hash-set here.

Reply:
Good catch, I'll fix it.